src/Evolution/Actions/AddMeshVelocityNonconservative.hpp

```diff
@@ -0,0 +1,104 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#pragma once
+
+#include <boost/optional.hpp>
+#include <tuple>
+
+#include "DataStructures/DataBox/DataBox.hpp"
+#include "DataStructures/DataBox/Prefixes.hpp"
+#include "Domain/TagsTimeDependent.hpp"
+#include "NumericalAlgorithms/LinearOperators/PartialDerivatives.hpp"
+#include "Utilities/Gsl.hpp"
+#include "Utilities/Requires.hpp"
+#include "Utilities/TMPL.hpp"
+#include "Utilities/TaggedTuple.hpp"
+
+/// \cond
+namespace Frame {
+struct Inertial;
+}  // namespace Frame
+namespace Parallel {
+template <typename Metavariables>
+class ConstGlobalCache;
+}  // namespace Parallel
+// IWYU pragma: no_forward_declare db::DataBox
+/// \endcond
+
+namespace evolution {
+namespace Actions {
+/*!
+ * \ingroup ActionsGroup
+ * \brief Compute and add the advective term for nonconservative systems on
+ * moving meshes
+ *
+ * Adds the following term to the time derivative:
+ *
+ * \f{align}{
+ *  v^i_g \partial_i u_\alpha,
+ * \f}
+ *
+ * where \f$u_\alpha\f$ are the evolved variables and \f$v^i_g\f$ is the
+ * velocity of the mesh.
+ *
+ * Uses:
+ * - DataBox:
+ *   - `Tags::deriv<System::variables_tags, Dim, Frame::Inertial>`
+ *   - `domain::Tags::MeshVelocity<Dim>`
+ *
+ * DataBox changes:
+ * - Adds: nothing
+ * - Removes: nothing
+ * - Modifies: `Tags::dt<system::variable_tags>`
+ */
+struct AddMeshVelocityNonconservative {
+  template <typename DbTagsList, typename... InboxTags, typename Metavariables,
+            typename ArrayIndex, typename ActionList,
+            typename ParallelComponent>
+  static std::tuple<db::DataBox<DbTagsList>&&> apply(
+      db::DataBox<DbTagsList>& box,
+      const tuples::TaggedTuple<InboxTags...>& /*inboxes*/,
+      const Parallel::ConstGlobalCache<Metavariables>& /*cache*/,
+      const ArrayIndex& /*array_index*/, ActionList /*meta*/,
+      const ParallelComponent* const /*meta*/) noexcept {  // NOLINT const
+    using gradients_tags = typename Metavariables::system::gradients_tags;
```

Comment:
From what I understand from `ScalarWave`, `system::gradients_tags` is a subset of the variables you require to take the gradient of in order to solve the equations on a time-independent grid. So, for a  different linear system such as  

\partial_t u_1 + \partial_x u_2 = 0
\partial_t u_2 + \partial_x u_2 = 0,

I'd have `gradients_tags = tmpl::list<U2>`. If this is correct, then line 73 would give `dt_gradient_tag = tmpl::list<Tags::dt<U2>>`, so line 79 would only mutate the equation for u_2, assuming that `dt_var_ptr` is passed with the right type. This, of course, might not be the case, as the type might actually be `tmpl::list<Tags::dt<U1>, Tags::dt<U2>>`, as you implicitly do in the test, so the code might not compile. 

What's wrong with my reasoning?

Reply:
That's a great question! The nonconservative system is written in the general form:

d_t u_{\alpha} + B^i_{\alpha \beta} d_i u_{\beta} - v^i_g d_i u_{\beta} = S_{\alpha}

So you're right, this should be using `variables_tag`. I've expanded the test to catch the bug.

Reply:
OK thanks for explaining this. Yeah, I was wondering why `variables_tag` wasn't being used here... 

Side comment: I guess the sign of the term proportional to the mesh velocity can be plus or minus depending on which frame you evaluate the equation in. Maybe an extra comment might be helpful, unless you're planning on documenting all this elsewhere.

Reply:
The sign is `+` in code is because it's brought over to the RHS, so

d_t u_{\alpha} = - B^i_{\alpha \beta} d_i u_{\beta} + v^i_g d_i u_{\beta}  + S_{\alpha}

The plan is to always evaluate the terms in the "inertial" frame because that's the easiest way to understand the equations and formulation. Where do you suggest I clarify the sign? In Doxygen or a C++ comment?

Reply:
OK yeah inertial would have been my guess. A Doxygen \note in the dox should suffice. Thanks!

Reply:
Okay, added a comment and squashed. Please let me know whether or not it is clear :)

Reply:
Looks clear to me, thanks.